crates/error/src/error.rs

```diff
@@ -0,0 +1,1812 @@
+        debug_assert!(self.is_oom());
+        let dangling = NonNull::<OutOfMemory>::dangling();
+        debug_assert_eq!(mem::size_of::<OutOfMemory>(), 0);
+        // Safety: it is always valid to turn `T`'s dangling pointer into an
+        // `&T` reference for unit types.
+        unsafe { dangling.as_ref() }
+    }
+
+    /// # Safety
+    ///
+    /// `self.is_oom()` must be true.
+    unsafe fn unchecked_oom_mut(&mut self) -> &mut OutOfMemory {
+        debug_assert!(self.is_oom());
+        let mut dangling = NonNull::<OutOfMemory>::dangling();
+        debug_assert_eq!(mem::size_of::<OutOfMemory>(), 0);
+        // Safety: it is always valid to turn `T`'s dangling pointer into an
+        // `&mut T` reference for unit types.
+        unsafe { dangling.as_mut() }
+    }
+
+    /// # Safety
+    ///
+    /// `self.is_boxed_dyn_error()` must be true.
+    unsafe fn unchecked_into_dyn_error(self) -> OwnedPtr<DynError> {
+        debug_assert!(self.is_boxed_dyn_error());
+        let inner = self.inner.cast::<DynError>();
+        mem::forget(self);
+        OwnedPtr::new(inner)
+    }
+
+    /// # Safety
+    ///
+    /// `self.is_boxed_dyn_error()` must be true.
+    unsafe fn unchecked_dyn_error_ref(&self) -> SharedPtr<'_, DynError> {
+        debug_assert!(self.is_boxed_dyn_error());
+        SharedPtr::new(self.inner.cast::<DynError>())
+    }
+
+    /// # Safety
+    ///
+    /// `self.is_boxed_dyn_error()` must be true.
+    unsafe fn unchecked_dyn_error_mut(&mut self) -> MutPtr<'_, DynError> {
+        debug_assert!(self.is_boxed_dyn_error());
+        MutPtr::new(self.inner.cast::<DynError>())
+    }
+
+    pub(crate) fn unpack(&self) -> OomOrDynErrorRef<'_> {
+        if self.is_oom() {
+            // Safety: is_oom() is true.
+            OomOrDynErrorRef::Oom(unsafe { self.unchecked_oom() })
+        } else {
+            debug_assert!(self.is_boxed_dyn_error());
+            // Safety: self.is_boxed_dyn_error() is true.
+            OomOrDynErrorRef::DynError(unsafe { self.unchecked_dyn_error_ref() })
+        }
+    }
+
+    pub(crate) fn unpack_mut(&mut self) -> OomOrDynErrorMut<'_> {
+        if self.is_oom() {
+            // Safety: self.is_oom() is true
+            OomOrDynErrorMut::Oom(unsafe { self.unchecked_oom_mut() })
+        } else {
+            debug_assert!(self.is_boxed_dyn_error());
+            // Safety: self.is_boxed_dyn_error() is true.
+            OomOrDynErrorMut::DynError(unsafe { self.unchecked_dyn_error_mut() })
+        }
+    }
+
+    pub(crate) fn into_boxed_dyn_core_error(
+        self,
+    ) -> Box<dyn core::error::Error + Send + Sync + 'static> {
+        let box_dyn_error_of_oom = || {
+            let ptr = NonNull::<OutOfMemory>::dangling().as_ptr();
+            // Safety: it is always safe to call `Box::<T>::from_raw` on `T`'s
+            // dangling pointer if `T` is a unit type.
+            let boxed = unsafe { Box::from_raw(ptr) };
+            boxed as _
```

Comment:
Instead of using `unsafe` could this do `Box::new(OutOfMemory::new())`? That won't actually allocate anything for zero-sized-types